crypto/fipsmodule/service_indicator/service_indicator_test.cc

```diff
@@ -676,6 +671,317 @@ TEST_P(HMAC_ServiceIndicatorTest, HMACTest) {
+ ASSERT_FALSE(RSA_generate_key_fips(rsa.get(), 4097, nullptr));
+
+ // Test that we can generate keys of the supported lengths:
+ for (const size_t bits : {2048, 3072, 4096}) {
+   SCOPED_TRACE(bits);
+
+   rsa.reset(RSA_new());
+   CALL_SERVICE_AND_CHECK_APPROVED(approved,
+               ASSERT_TRUE(RSA_generate_key_fips(rsa.get(), bits, nullptr)));
+   ASSERT_EQ(approved, AWSLC_APPROVED);
+   ASSERT_EQ(bits, BN_num_bits(rsa->n));
+ }
+}
+
+struct RSATestVector {
+  // key_size is the input rsa key size.
+  const int key_size;
+  // md_func is the digest to test.
+  const EVP_MD *(*func)(void);
+  // expected to be approved or not for signature generation.
+  const int sig_gen_expect_approved;
+  // expected to be approved or not for signature verification.
+  const int sig_ver_expect_approved;
+};
+struct RSATestVector kRSASigGenTestVectors[] = {
+    { 1024, &EVP_sha1, AWSLC_NOT_APPROVED, AWSLC_APPROVED },
+    { 1024, &EVP_sha256, AWSLC_NOT_APPROVED, AWSLC_APPROVED },
+    { 1024, &EVP_sha512, AWSLC_NOT_APPROVED, AWSLC_APPROVED },
+
+    { 2048, &EVP_sha1, AWSLC_NOT_APPROVED, AWSLC_APPROVED },
+    { 2048, &EVP_sha224, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 2048, &EVP_sha256, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 2048, &EVP_sha384, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 2048, &EVP_sha512, AWSLC_APPROVED, AWSLC_APPROVED },
+
+    { 3072, &EVP_sha1, AWSLC_NOT_APPROVED, AWSLC_APPROVED },
+    { 3072, &EVP_sha224, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 3072, &EVP_sha256, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 3072, &EVP_sha384, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 3072, &EVP_sha512, AWSLC_APPROVED, AWSLC_APPROVED },
+
+    { 4096, &EVP_sha1, AWSLC_NOT_APPROVED, AWSLC_APPROVED },
+    { 4096, &EVP_sha224, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 4096, &EVP_sha256, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 4096, &EVP_sha384, AWSLC_APPROVED, AWSLC_APPROVED },
+    { 4096, &EVP_sha512, AWSLC_APPROVED, AWSLC_APPROVED },
+};
+
+class RSA_ServiceIndicatorTest : public testing::TestWithParam<RSATestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, RSA_ServiceIndicatorTest, testing::ValuesIn(kRSASigGenTestVectors));
+
+TEST_P(RSA_ServiceIndicatorTest, RSASigGen) {
+  const RSATestVector &rsaTestVector = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+
+  bssl::UniquePtr<RSA> rsa(RSA_new());
+  bssl::UniquePtr<BIGNUM> e(BN_new());
+  bssl::UniquePtr<EVP_PKEY> pkey(EVP_PKEY_new());
+  bssl::ScopedEVP_MD_CTX md_ctx;
+  ASSERT_TRUE(pkey);
+  ASSERT_TRUE(rsa);
+  BN_set_word(e.get(), RSA_F4);
+
+  // Generate a generic rsa key.
+  RSA_generate_key_ex(rsa.get(), rsaTestVector.key_size, e.get(),
+                      nullptr);
+  ASSERT_TRUE(EVP_PKEY_set1_RSA(pkey.get(), rsa.get()));
+
+  // Test running the EVP_DigestSign interfaces one by one directly, and check
+  // |EVP_DigestSignFinal| for approval at the end.
+  std::vector<uint8_t> final_output;
+  size_t sig_len;
+  ASSERT_TRUE(EVP_DigestSignInit(
```

Comment:
I think in previous tests, you checked that these APIs set the SI to `NOT_APPROVED`.